livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py

```diff
@@ -302,23 +304,67 @@ async def recv_task():
 
         await asyncio.gather(send_task(), recv_task(), keepalive_task())
 
+    def _commit_all_final_events(self) -> None:
+        if not self._speaking:
+            logging.warning(
+                "trying to commit final events without being in the speaking state"
+            )
+            return
+
+        if len(self._final_events) == 0:
+            logging.warning("received end of speech without any final transcription")
+
+        # combine all final transcripts since the start of the speech
+        sentence = ""
+        confidence = 0
+        for alt in self._final_events:
+            sentence += f"{alt.alternatives[0].text.strip()} "
+            confidence += alt.alternatives[0].confidence
+
+        sentence = sentence.rstrip()
+        confidence /= len(self._final_events)  # avg. of confidence
+
+        self._speaking = False
+        end_event = stt.SpeechEvent(
+            type=stt.SpeechEventType.END_OF_SPEECH,
+            alternatives=[
+                stt.SpeechData(
+                    language=self._config.language,
+                    start_time=self._final_events[0].alternatives[0].start_time,
+                    end_time=self._final_events[-1].alternatives[0].end_time,
+                    confidence=confidence,
+                    text=sentence,
+                )
+            ],
+        )
+        self._event_queue.put_nowait(end_event)
+        self._final_events = []
+
     def _process_stream_event(self, data: dict) -> None:
         assert self._config.language is not None
 
-        # https://developers.deepgram.com/docs/speech-started
         if data["type"] == "SpeechStarted":
+            if self._speaking:
+                logging.warning("received SpeechStarted while already speaking")
+
+            self._speaking = True
             start_event = stt.SpeechEvent(type=stt.SpeechEventType.START_OF_SPEECH)
             self._event_queue.put_nowait(start_event)
-            return
+        elif data["type"] == "UtteranceEnd":
+            # https://developers.deepgram.com/docs/understanding-end-of-speech-detection#using-utteranceend-and-endpointing
+            # If UtterenceEnd comes before SpeechFinal, we use it to commit the final events
+            # otherwise, we ignore it.
+            if self._speaking:
+                self._commit_all_final_events()
```

Comment:
Can UtteranceEnd happens without SpeechStart?

Reply:
Shouldn't, unless you receive transcribed words without a speech start. Utterance end is based on transcribed words timing:
https://developers.deepgram.com/docs/understanding-end-of-speech-detection#using-utteranceend